chips/nrf52/src/qdec.rs

```diff
@@ -0,0 +1,285 @@
+        REPORTRDY_READCLRACC 0,
+        /// Write '1' to Enable shortcut on EVENTS_COMPARE\[1\] event
+        SAMPLERDY_STOP 1,
+        /// Write '1' to Enable shortcut on EVENTS_COMPARE\[2\] event
+        REPORTRDY_RDCLRACC 2,
+        /// Write '1' to Enable shortcut on EVENTS_COMPARE\[3\] event
+        REPORTRDY_STOP 3,
+        /// Write '1' to Enable shortcut on EVENTS_COMPARE\[4\] event
+        DBLRDY_RDCLRDBL 4,
+        /// Write '1' to Enable shortcut on EVENTS_COMPARE\[5\] event
+        DBLRDY_STOP 5,
+        /// Write '1' to Enable shortcut on EVENTS_COMPARE\[6\] event
+        SAMPLERDY_READCLRACC 6
+    ],
+    Event [
+        READY 0
+    ],
+    PinSelect [
+        Pin OFFSET(0) NUMBITS(5),
+        Port OFFSET(5) NUMBITS(1),
+        Connect OFFSET(31) NUMBITS(1)
+    ],
+    Inte [
+        /// Write '1' to Enable interrupt on EVENTS_COMPARE\[0\] event
+        SAMPLERDY 0,
+        /// Write '1' to Enable interrupt on EVENTS_COMPARE\[1\] event
+        REPORTRDY 1,
+        /// Write '1' to Enable interrupt on EVENTS_COMPARE\[2\] event
+        ACCOF 2,
+        /// Write '1' to Enable interrupt on EVENTS_COMPARE\[3\] event
+        DBLRDY 3,
+        /// Write '1' to Enable interrupt on EVENTS_COMPARE\[4\] event
+        STOPPED 4
+    ],
+    LedPol [
+        LedPol OFFSET(0) NUMBITS(1) [
+            ActiveLow = 0,
+            ActiveHigh = 1
+        ]
+    ],
+    Sample [
+        SAMPLE 1
+    ],
+    SampPer [
+        SAMPLEPER OFFSET(0) NUMBITS(4) [
+            us128 = 0,
+            us256 = 1,
+            us512 = 2,
+            us1024 = 3,
+            us2048 = 4,
+            us4096 = 5,
+            us8192 = 6,
+            us16384 = 7,
+            ms32 = 8,
+            ms65 = 9,
+            ms131 = 10
+        ]
+    ],
+    ReportPer [
+        REPORTPER OFFSET(0) NUMBITS(4) [
+            hz10 = 0,
+            hz40 = 1,
+            hz80 = 2,
+            hz120 = 3,
+            hz160 = 4,
+            hz200 = 5,
+            hz240 = 6,
+            hz280 = 7,
+            hz1 = 8
+        ]
+    ],
+    Acc [
+        ACC OFFSET(0) NUMBITS(32)
+    ],
+    AccDbl [
+        ACCDBL OFFSET(0) NUMBITS(4)
+    ]
+];
+
+/// This defines the beginning of memory which is memory-mapped to the Qdec
+/// This base is declared under the Registers Table 3
+const QDEC_BASE: StaticRef<QdecRegisters> =
+    unsafe { StaticRef::new(0x40012000 as *const QdecRegisters) };
+
+pub static mut QDEC: Qdec = Qdec::new();
+
+/// Qdec type declaration: gives the Qdec instance registers and a client
+pub struct Qdec {
+    registers: StaticRef<QdecRegisters>,
+    client: OptionalCell<&'static dyn kernel::hil::qdec::QdecClient>,
+    state: usize,
+}
+
+/// Qdec impl: provides the Qdec type with vital functionality including:
+impl Qdec {
+    const fn new() -> Qdec {
+        let qdec = Qdec {
+            registers: QDEC_BASE,
+            client: OptionalCell::empty(),
+            state: 0, //Start
```

Comment:
This state should be an enum describing the possible states of the state machine, rather than a usize

Reply:
added an enum in there: QdecState, which has two settings right now: Start and Stop